go/externals/proof_service_generic_social.go

```diff
@@ -0,0 +1,158 @@
+}
+
+//
+//=============================================================================
+
+// Validated configuration from the server
+type GenericSocialProofConfig struct {
+	keybase1.ParamProofServiceConfig
+	usernameRe *regexp.Regexp
+}
+
+func NewGenericSocialProofConfig(config keybase1.ParamProofServiceConfig) (*GenericSocialProofConfig, error) {
+	gsConfig := &GenericSocialProofConfig{
+		ParamProofServiceConfig: config,
+		usernameRe:              nil,
+	}
+	err := gsConfig.validate()
+	if err != nil {
+		return nil, err
+	}
+	return gsConfig, nil
+}
+
+func (c *GenericSocialProofConfig) validate() error {
+	re, err := regexp.Compile(c.Username.Re)
+	if err != nil {
+		return err
+	}
+	c.usernameRe = re
+	if !strings.Contains(c.PrefillUrl, kbUsernameKey) {
+		return fmt.Errorf("invalid PrefillUrl %s, missing %s", c.PrefillUrl, kbUsernameKey)
+	}
+	if !strings.Contains(c.PrefillUrl, sigHashKey) {
+		return fmt.Errorf("invalid PrefillUrl %s, missing %s", c.PrefillUrl, sigHashKey)
+	}
+	if !strings.Contains(c.CheckUrl, usernameKey) {
+		return fmt.Errorf("invalid CheckUrl %s, missing %s", c.CheckUrl, usernameKey)
+	}
+	return nil
+}
+
+func (c *GenericSocialProofConfig) prefillURLWithValues(kbUsername string, sigID keybase1.SigID) (string, error) {
+	url := strings.Replace(c.PrefillUrl, kbUsernameKey, kbUsername, 1)
+	url = strings.Replace(url, sigHashKey, sigID.String(), 1)
+	if !strings.Contains(url, kbUsername) {
+		return "", fmt.Errorf("Invalid PrefillUrl, missing kbUsername")
+	}
+	if !strings.Contains(url, sigID.String()) {
+		return "", fmt.Errorf("Invalid PrefillUrl, missing sigHash")
+	}
+	return url, nil
+}
+
+//=============================================================================
+
+type GenericSocialProofServiceType struct {
+	libkb.BaseServiceType
+	config *GenericSocialProofConfig
+}
+
+func NewGenericSocialProofServiceType(config *GenericSocialProofConfig) GenericSocialProofServiceType {
+	return GenericSocialProofServiceType{
+		config: config,
+	}
+}
+
+func (t GenericSocialProofServiceType) AllStringKeys() []string { return t.BaseAllStringKeys(t) }
+
+func (t GenericSocialProofServiceType) NormalizeUsername(s string) (string, error) {
+	if !t.config.usernameRe.MatchString(s) {
+		return "", libkb.NewBadUsernameError(s)
+	}
+	// TODO always normalize ToLower?
+	return strings.ToLower(s), nil
+}
+
+func (t GenericSocialProofServiceType) NormalizeRemoteName(m libkb.MetaContext, s string) (ret string, err error) {
+	return t.NormalizeUsername(s)
+}
+
+func (t GenericSocialProofServiceType) GetPrompt() string {
+	return fmt.Sprintf("Your username on %s", t.config.DisplayName)
+}
+
+func (t GenericSocialProofServiceType) ToServiceJSON(username string) *jsonw.Wrapper {
+	return t.BaseToServiceJSON(t, username)
+}
+
+func (t GenericSocialProofServiceType) PostInstructions(username string) *libkb.Markup {
+	return libkb.FmtMarkup(`Please click on the following link to post to %v:`, t.config.DisplayName)
```

Comment:
Putting the link on its own line might be nicer for CLI

Reply:
kk will split in downstream PR

Reply:
this was already on one line, just forgot to update the description text: https://github.com/keybase/client/pull/13832/files#diff-9507f7d55a600de3f61e22495419b9d5R148